### ISSUE TITLE
'iOS-webkit-debug-proxy' crash happens when test on real device

### ISSUE THREAD

**Issue by scnymc** [appium/appium#4198]
Hi,

iOS-webkit-debug-proxy crashes when run the Appium for Safari on real device. is it a known bug? if yes, any solutions for it?

Here is my test machine configurations:
1. Mac os: 10.10; xCode: 6.1
2. Appium 1.3.3 (install using npm)
3. SafariLauncher, I got the source from github and sign it with my developer signature, the bundle ID is "com.bytearc.SafariLauncher", and has been installed to my iPad
4. iOS-webkit-debug-proxy is started with the command "Ios_webkit_debug_proxy -c 9273bc6013c7645d05addf2ae7733c7b616e8790:27753"  (its version is 1.4)
5. My capabilities:
capabilities.setCapability("platformName", "iOS");
capabilities.setCapability("platformVersion", "8.1");
capabilities.setCapability("deviceName", "iPad");
capabilities.setCapability("udid","xxxxx");
capabilities.setCapability(MobileCapabilityType.BROWSER_NAME, "safari");
driver = new AppiumDriver(new URL("http://127.0.0.1:4723/wd/hub"),capabilities);

When I run the java code, the safariLauncher is launched on device, but the ios_webkit_debug_proxy is crashed! the log as below:
=====================
macsbjlhl:~ yuemingchen$ ios_webkit_debug_proxy -c 9273bc6013cMYIPADID733c7b616e8790:27753
Connected :27753 to BTG iPad (9273bc6013c7645d05addf2ae7733c7b616e8790)
Invalid message _rpc_applicationUpdated: <dict>
	<key>WIRApplicationBundleIdentifierKey</key>
	<string>com.apple.mobilesafari</string>
	<key>WIRApplicationNameKey</key>
	<string>Safari</string>
	<key>WIRIsApplicationProxyKey</key>
	<false/>
	<key>WIRIsApplicationActiveKey</key>
	<integer>1</integer>
	<key>WIRApplicationIdentifierKey</key>
	<string>PID:2353</string>
</dict>
Disconnected :27753 from BTG iPad (9273bc6013c7MYIPADID7733c7b616e8790)
===================

And here is the Appium log:
===================
macsbjlhl:~ yuemingchen$ appium -U 9273bc6013c76mypadeID7733c7b616e8790
info: Welcome to Appium v1.3.3 (REV 34e21b01b87d4e3959e1c97c0a8d0b1a610bdc78)
info: Appium REST http interface listener started on 0.0.0.0:4723
info: [debug] Non-default server args: {"udid":"9273bc6013c7645dMYPADae7733c7b616e8790"}
info: Console LogLevel: debug
info: --> POST /wd/hub/session {"desiredCapabilities":{"automationName":null,"orientation":"LANDSCAPE","app":null,"platformVersion":"8.1","deviceName":"iPad","platformName":"iOS","browserName":"safari","appActivity":null,"unicodeKeyboard":null,"udid":"9273bc6013c7645d05aMYPADe7733c7b616e8790","appPackage":null}}
info: Client User-Agent string: Apache-HttpClient/4.3.3 (java 1.5)
info: [debug] Configuring Safari session
info: [debug] Using local .zip from command line: /usr/local/lib/node_modules/appium/build/SafariLauncher/SafariLauncher.zip
info: [debug] Copying local zip to tmp dir
info: [debug] /usr/local/lib/node_modules/appium/build/SafariLauncher/SafariLauncher.zip copied to /var/folders/b1/z0t1mqjx63n58k54gq82sdnw0000gr/T/114111-5867-1dznbch/appium-app.zip
info: [debug] Unzipping /var/folders/b1/z0t1mqjx63n58k54gq82sdnw0000gr/T/114111-5867-1dznbch/appium-app.zip
info: [debug] Testing zip archive: /var/folders/b1/z0t1mqjx63n58k54gq82sdnw0000gr/T/114111-5867-1dznbch/appium-app.zip
info: [debug] Zip archive tested clean
info: [debug] Unzip successful
info: [debug] Using locally extracted app: /var/folders/b1/z0t1mqjx63n58k54gq82sdnw0000gr/T/114111-5867-1dznbch/submodules/SafariLauncher/build/Release-iphoneos/SafariLauncher.app
info: [debug] Creating new appium session d1710b2b-8475-4521-b70b-45504922b442
info: [debug] Removing any remaining instruments sockets
info: [debug] Cleaned up instruments socket /tmp/instruments_sock
info: [debug] Setting Xcode folder
info: [debug] Setting Xcode version
info: [debug] Setting iOS SDK Version
info: [debug] Getting sdk version from xcrun with a timeout
info: [debug] iOS SDK Version set to 8.1
info: [debug] Not checking whether simulator is available since we're on a real device
info: [debug] Detecting automation tracetemplate
info: [debug] Not auto-detecting udid, running on sim
info: [debug] Could not parse plist file (as binary) at /var/folders/b1/z0t1mqjx63n58k54gq82sdnw0000gr/T/114111-5867-1dznbch/submodules/SafariLauncher/build/Release-iphoneos/SafariLauncher.app/en.lproj/Localizable.strings
info: Will try to parse the plist file as XML
`parseFileSync()` is deprecated. Use `parseStringSync()` instead.
info: [debug] Could not parse plist file (as XML) at /var/folders/b1/z0t1mqjx63n58k54gq82sdnw0000gr/T/114111-5867-1dznbch/submodules/SafariLauncher/build/Release-iphoneos/SafariLauncher.app/en.lproj/Localizable.strings
warn: Could not parse app Localizable.strings; assuming it doesn't exist
info: [debug] Getting bundle ID from app
info: [debug] Parsed app Info.plist (as binary)
info: [debug] Creating instruments
info: On some xcode 6 platforms, instruments-without-delay does not work. If you experience this, you will need to re-run appium with the --native-instruments-lib flag
info: [debug] Preparing uiauto bootstrap
info: [debug] Dynamic bootstrap dir: /Users/yuemingchen/Library/Application Support/appium/bootstrap
info: [debug] Dynamic env: {"nodePath":"/usr/local/bin/node","commandProxyClientPath":"/usr/local/lib/node_modules/appium/node_modules/appium-uiauto/bin/command-proxy-client.js","instrumentsSock":"/tmp/instruments_sock","interKeyDelay":null,"justLoopInfinitely":false,"autoAcceptAlerts":false,"sendKeyStrategy":"grouped"}
info: [debug] Dynamic bootstrap code: // This file is automatically generated. D...ot manually modify!
info: [debug] Dynamic bootstrap path: /Users/yuemingchen/Library/Application Support/appium/bootstrap/bootstrap-d0f21a5b61bb08af.js
info: [debug] Reusing dynamic bootstrap: /Users/yuemingchen/Library/Application Support/appium/bootstrap/bootstrap-d0f21a5b61bb08af.js
info: [debug] Getting device string from opts: {"forceIphone":false,"forceIpad":false,"xcodeVersion":"6.1","iOSSDKVersion":"8.1","deviceName":"iPad","platformVersion":"8.1"}
info: [debug] fixDevice is on
info: [debug] Final device string is: 'iPad (8.1 Simulator)'
info: [debug] Not setting device type since we're on a real device
info: [debug] Checking whether we need to set app preferences
info: [debug] Not setting iOS and app preferences since we're on a real device
info: [debug] Running ios sim reset flow
info: [debug] Killing the simulator process
info: [debug] Killall iOS Simulator
info: [debug] Killing any other simulator daemons
info: [debug] On a real device; cannot clean device state
info: [debug] Not setting locale because we're using a real device
info: [debug] No iOS / app preferences to set
info: [debug] Starting iOS device log capture via deviceconsole
info: [debug] Not pre-launching simulator
info: [debug] Checking app install status using: /usr/local/lib/node_modules/appium/build/fruitstrap/fruitstrap isInstalled --id 9273bc6013c7645d05addMYPAD733c7b616e8790 --bundle com.bytearc.SafariLauncher
info: [debug] Starting command proxy.
info: [debug] Instruments socket server started at /tmp/instruments_sock
info: [debug] Starting instruments
info: [debug] Instruments is at: /Applications/Xcode.app/Contents/Developer/usr/bin/instruments
info: Launching instruments
info: [debug] Attempting to run app on real device with UDID 9273bc6013c7645dMYPADID3c7b616e8790
info: [debug] Spawning instruments with command: /Applications/Xcode.app/Contents/Developer/usr/bin/instruments -t /Applications/Xcode.app/Contents/Applications/Instruments.app/Contents/PlugIns/AutomationInstrument.xrplugin/Contents/Resources/Automation.tracetemplate -D /tmp/appium-instruments/instrumentscli0.trace -w 9273bc6013c7645d05MYPADID2ae7733c7b616e8790 com.bytearc.SafariLauncher -e UIASCRIPT "/Users/yuemingchen/Library/Application Support/appium/bootstrap/bootstrap-d0f21a5b61bb08af.js" -e UIARESULTSPATH /tmp/appium-instruments
info: [debug] And extra without-delay env: {}
info: [debug] And launch timeouts (in ms): {"global":90000}
info: [debug] [INST STDERR] 2014-12-01 15:47:09.177 instruments[5900:136619] WebKit Threading Violation - initial use of WebKit from a secondary thread.
info: Instruments is ready to receive commands
info: [debug] Instruments launched. Starting poll loop for new commands.
info: [debug] Setting bootstrap config keys/values
info: [debug] Pushing command to appium work queue: "target = $.target();\nau = $;\n$.isVerbose = true;\n"
info: [debug] Socket data received (2 bytes)
info: [debug] Socket data being routed.
info: [debug] Sending command to instruments: target = $.target();
au = $;
$.isVerbose = true;

info: [debug] [INST] 2014-12-01 07:45:27 +0000 Debug: evaluation finished
info: [debug] [INST] 2014-12-01 07:45:27 +0000 Debug: responding with:
info: [debug] [INST] 2014-12-01 07:45:27 +0000 Debug: Running system command #1: /usr/local/bin/node /usr/local/lib/node_modules/appium/node_modules/appium-uiauto/bin/command-proxy-client.js /tmp/instruments_sock 2,{"status":0,"value":true}...
info: [debug] Socket data received (27 bytes)
info: [debug] Socket data being routed.
info: [debug] Got result from instruments: {"status":0,"value":true}
info: [debug] Setting initial orientation to LANDSCAPE
info: [debug] Pushing command to appium work queue: "au.setScreenOrientation('LANDSCAPE')"
info: [debug] Sending command to instruments: au.setScreenOrientation('LANDSCAPE')
info: [debug] [INST] 2014-12-01 07:45:28 +0000 Debug: Got new command 1 from instruments: au.setScreenOrientation('LANDSCAPE')
info: [debug] [INST] 2014-12-01 07:45:28 +0000 Debug: evaluating au.setScreenOrientation('LANDSCAPE')
info: [debug] [INST] 2014-12-01 07:45:28 +0000 Debug: target.setDeviceOrientation("3")
info: [debug] [INST] 2014-12-01 07:45:29 +0000 Debug: evaluation finished
info: [debug] Socket data received (34 bytes)
info: [debug] Socket data being routed.
info: [debug] Got result from instruments: {"status":0,"value":"LANDSCAPE"}
info: [debug] Waiting for app source to contain elements
info: [debug] Pushing command to appium work queue: "au.mainApp().getTreeForXML()"
info: [debug] Sending command to instruments: au.mainApp().getTreeForXML()
info: [debug] [INST] 2014-12-01 07:45:29 +0000 Debug: responding with:
info: [debug] [INST] 2014-12-01 07:45:29 +0000 Debug: Running system command #2: /usr/local/bin/node /usr/local/lib/node_modules/appium/node_modules/appium-uiauto/bin/command-proxy-client.js /tmp/instruments_sock 2,{"status":0,"value":"LANDSCAPE"}...
info: [debug] [INST] 2014-12-01 07:45:30 +0000 Debug: Got new command 2 from instruments: au.mainApp().getTreeForXML()
info: [debug] [INST] 2014-12-01 07:45:30 +0000 Debug: evaluating au.mainApp().getTreeForXML()
info: [debug] Socket data received (8100 bytes)
info: [debug] Socket data received (8168 bytes)
info: [debug] Socket data received (8056 bytes)
info: [debug] Socket data received (7261 bytes)
info: [debug] Socket data received (7609 bytes)
info: [debug] Socket data received (7363 bytes)
info: [debug] Socket data being routed.
info: [debug] Got result from instruments: {"status":0,"value":{"UIAApplication":{"@":{"name":"Safari","label":"Safari","value":null,"dom":null,"enabled":true,"valid":true,"visible":true,"hint":null,"path":"/0","x":20,"y":-256,"width":748,"height":1024},">":[{"UIAWindow":{"@":{"name":null,"label":null,"value":null,"dom":null,"enabled":true,"

info: [debug] [INST] 2014-12-01 07:45:41 +0000 Debug: evaluation finished
info: [debug] [INST] 2014-12-01 07:45:41 +0000 Debug: responding info: [debug] [INST] 2014-12-01 07:45:41 +0000 Debug: Running system command #3: /usr/local/bin/node /usr/local/lib/node_modules/appium/node_modules/appium-uiauto/bin/command-proxy-client.js /tmp/instruments_sock 2,{"status":0,"value":{"UIAApplication":{"@":{"name":"Safari","label":"Safari","value":null,"dom":null,"enabled":true,"valid":true,"visible":true,"hint":null,"path":"/0...
…(here I delete some page info)

info: [debug] Navigating to most recently opened webview
info: [debug] [REMOTE] Getting WebKitRemoteDebugger pageArray
error: [REMOTE] Could not connect to WebKitRemoteDebugger server
info: [debug] Cleaning up appium session
error: Failed to start an Appium session, err was: Error: Could not navigate to webview! Err: Error: connect ECONNREFUSED
info: [debug] Error: Could not navigate to webview! Err: Error: connect ECONNREFUSED
    at null.<anonymous> (/usr/local/lib/node_modules/appium/lib/devices/ios/ios.js:1556:12)
    at null.<anonymous> (/usr/local/lib/node_modules/appium/lib/devices/ios/ios-controller.js:1532:14)
    at null.<anonymous> (/usr/local/lib/node_modules/appium/lib/devices/ios/webkit-remote-debugger.js:73:14)
    at ClientRequest.<anonymous> (/usr/local/lib/node_modules/appium/node_modules/underscore/underscore.js:754:19)
    at ClientRequest.emit (events.js:95:17)
    at Socket.socketErrorListener (http.js:1551:9)
    at Socket.emit (events.js:95:17)
    at net.js:440:14
    at process._tickDomainCallback (node.js:463:13)
info: [debug] Responding to client with error: {"status":33,"value":{"message":"A new session could not be created. (Original error: Could not navigate to webview! Err: Error: connect ECONNREFUSED)","origValue":"Could not navigate to webview! Err: Error: connect ECONNREFUSED"},"sessionId":null}
info: <-- POST /wd/hub/session 500 26107.103 ms - 248

**Comment by amarnathrayudu** [comment]
@scnymc 
I had the same issue with ios 8.1.3. I've removed the SafariLauncher from device and installed with fruitstrap again.
Then started ios-webkit-debug-proxy-launcher.js and faced the same issue for some time, I've disconnected the device and connected again ... it worked like charm.

Thanks,
Amarnath

**Comment by sujittarun** [comment]
@amarnathrayudu @scnymc I had stuck with same issue. 
1. How to install SafariLauncher with fruitstrap?
2. how to start ios-webkit-debug-proxy-launcher.js?

**Comment by 0x1mason** [comment]
Need to replace ios_webkit_debug_proxy_launcher.js workaround with iwdp that includes @jchuong fix.

**Comment by amarnathrayudu** [comment]
Hi @0x1mason, 
Is iwdp fix available in latest release ?
Hi @sujittarun,
If you have not resolved it, please look at it https://discuss.appium.io/t/safari-on-ios-ios-webkit-debug-crashing-when-safarilauncher-launches-safari/1208/16 . We have already dicussed about it.

Thanks,
Amarnath Rayudu

**Comment by moizjv** [comment]
@sujittarun [documentation](https://github.com/appium/appium/blob/master/docs/en/advanced-concepts/ios-webkit-debug-proxy.md) for ios-debug-proxy.

**Comment by Jonahss** [comment]
So everything should be settled here now. Anyone else still having issues? If so, comment here and we will reopen the issue.

**Comment by Sendami** [comment]
i've done all those steps and I'm still having that error

**Comment by NehaDua** [comment]
Hi All,

I was also facing this webkit crashing issue and trying its work around "./bin/ios-webkit-debug-proxy-launcher.js -c :27753 -d"

first i did "cd /usr/local/bin" on terminal
then "ios-webkit-debug-proxy-launcher.js -c ca666e295c7cd5a893e376989f3c4a3b24987534:27753 –d"

it says "command not found "

can someone please

**Comment by priyatech** [comment]
iOS Web Kit debug Proxy connects with the iOS device and disconnects immediately. Once in 10 times plugging and unplugging and changing the port it connects. Is there any  iOS Web Kit debug Proxy cache on device or mac to clean up? 

Xcode7.2 
OSx10.11
iOS Device 7.1
iPhone 5S

**Comment by luiscspinho** [comment]
I'm still facing this issue. The first time I run the test, everything looks fine, however if I run the same test again on the real device the following message appears and the debug proxy crashes:

> Invalid message _rpc_applicationUpdated: <dict>
> 	<key>WIRApplicationIdentifierKey</key>
> 	<string>PID:273</string>
> 	<key>WIRIsApplicationProxyKey</key>
> 	<true/>
> 	<key>WIRApplicationNameKey</key>
> 	<string></string>
> 	<key>WIRApplicationBundleIdentifierKey</key>
> 	<string>com.apple.WebKit.WebContent</string>
> 	<key>WIRIsApplicationActiveKey</key>
> 	<integer>0</integer>
> 	<key>WIRHostApplicationIdentifierKey</key>
> 	<string>PID:270</string>
> </dict>
> Disconnected :27753 from iPhone5s (66dea3d32384dbea9fae4f915e3274608c12a481)

If I manually close the safari launcher on the phone and restart the debug proxy, everything is well until the second time I run the test.

**Comment by imurchie** [comment]
Do you have the latest debug proxy?

**Comment by luiscspinho** [comment]
Yes, I do, i just installed it from brew today

**Comment by imurchie** [comment]
What are the details of your system? Appium version, node and npm versions, etc.

**Comment by luiscspinho** [comment]
Appium v1.5.1, node v0.12.9, npm v2.15.0, el capitan 10.11.4 running a iPhone 5s 9.1

**Comment by shivakumar89** [comment]
I am facing similar kind of issue, I started iwdp like this : ios_webkit_debug_proxy -c UDID:27755 -d and faced device disconnect error.

And I started using iwdp from appium bin : /bin/ios-webkit-debug-proxy-launcher.js -c udid:27753 -d it seems to work great where you cannot get disconnected any devices but sometimes when ur running multiple  tests its gets hung up and only can see apple.com in safari launcher. so then I have to close safari launcher or restart iwdp kit or use PM2 tool to restart it.
Logs:
[debug] [iOS] Executing iOS command 'setUrl'
[debug] [iOS] Attempting to set url 'my url'
[debug] [RemoteDebugger] Navigating to new URL: my url
[debug] [RemoteDebugger] Sending WebKit data: {"method":"Page.navigate","...
[debug] [RemoteDebugger] Receiving WebKit data: {"result":{},"id":1}
[debug] [RemoteDebugger] Found handler for message '1'
[debug] [RemoteDebugger] Received page navigated message: {"result":{},"id":1}
[debug] [RemoteDebugger] Waiting for dom...
[debug] [RemoteDebugger] Page loaded, verifying whether ready
[debug] [RemoteDebugger] Checking document readyState
[debug] [RemoteDebugger] Sending javascript command (function (){ return docume...
[debug] [RemoteDebugger] Sending WebKit data: {"method":"Runtime.evaluate...

So, can someone suggest me how to fix this ?

we need something like PM2 tool to keep devices alive all the time and run any scripts on all iOS devices.

Also is there any chance we can close safari after we finish tests using Appium. now I can close manually each time and run the tests again but is it possible to make it close each time with automated tests.

thanks in advance.
